lib/widgets/message_list.dart

```diff
@@ -725,12 +726,47 @@ class RecipientHeaderDate extends StatelessWidget {
           //   https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-caps#all-small-caps
           fontFeatures: const [FontFeature.enable('c2sc'), FontFeature.enable('smcp')],
         ).merge(weightVariableTextStyle(context)),
-        _kRecipientHeaderDateFormat.format(
-          DateTime.fromMillisecondsSinceEpoch(message.timestamp * 1000))));
+        formatHeaderDate(
+          zulipLocalizations,
+          DateTime.fromMillisecondsSinceEpoch(message.timestamp * 1000),
+          now: DateTime.now())));
   }
 }
 
-final _kRecipientHeaderDateFormat = DateFormat('y-MM-dd', 'en_US'); // TODO(#278)
+@visibleForTesting
+String formatHeaderDate(
+  ZulipLocalizations zulipLocalizations,
+  DateTime dateTime, {
+  required DateTime now,
+}) {
+  assert(!dateTime.isUtc && !dateTime.isUtc,
+    '`dateTime` and `now` need to be in local time.');
+  if (dateTime.year == now.year &&
```

Comment:
nit:
```suggestion
    '`dateTime` and `now` need to be in local time.');

  if (dateTime.year == now.year &&
```

The assert is less closely related to the same-day conditional than the latter is to what's after it. So since there's a blank line separating afterward, there should be a blank line separating before.